compiler/noirc_evaluator/src/ssa/opt/array_get.rs

```diff
@@ -0,0 +1,156 @@
+        let block = function.entry_block();
+        let dfg = &mut function.dfg;
+        let instructions = dfg[block].take_instructions();
+
+        for instruction_id in instructions {
+            let instruction = dfg[instruction_id].clone();
+
+            match &instruction {
+                Instruction::IfElse { then_value, .. } => {
+                    let then_value = *then_value;
+
+                    // Only apply this optimization to IfElse where values are arrays
+                    let Type::Array(..) = dfg.type_of_value(then_value) else {
+                        continue;
+                    };
+
+                    let results = dfg.instruction_results(instruction_id);
+                    let result = results[0];
+                    self.result_if_else.insert(result, instruction_id);
+
+                    dfg[block].instructions_mut().push(instruction_id);
+                }
+                Instruction::ArrayGet { array, index } => {
+                    // If this array get is for an array that is the result of a previous IfElse...
+                    if let Some(if_else) = self.result_if_else.get(array) {
+                        if let Instruction::IfElse {
+                            then_condition,
+                            then_value,
+                            else_condition,
+                            else_value,
+                            ..
+                        } = &dfg[*if_else]
+                        {
+                            let then_condition = *then_condition;
+                            let then_value = *then_value;
+                            let else_condition = *else_condition;
+                            let else_value = *else_value;
+
+                            let then_value_type = dfg.type_of_value(then_value);
+
+                            let Type::Array(element_type, _) = then_value_type else {
+                                panic!("ice: expected array type, got {:?}", then_value_type);
+                            };
+                            let element_type: &Vec<Type> = &element_type;
+
+                            // Given the original IfElse instruction is this:
+                            //
+                            //     v10 = if v0 then v2 else if v1 then v3
+                            //
+                            // and the ArrayGet instruction is this:
+                            //
+                            //     v11 = array_get v4, index v4
+
+                            // First create an instruction like this, for the then branch:
+                            //
+                            //     v12 = array_get v2, index v4
+                            let then_result = dfg.insert_instruction_and_results(
+                                Instruction::ArrayGet { array: then_value, index: *index },
+                                block,
+                                Some(element_type.clone()),
```

Comment:
Is there a way to avoid these clones?

Reply:
Cloning Types is fairly common and usually cheap.